src/helpers.rs

```diff
@@ -827,6 +881,123 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             None => tcx.item_name(def_id),
         }
     }
+
+    // extract the scalar value from the result of reading a scalar from the machine
+    // and convert it to a CArg
+    fn extract_scalar(
+        scalar_res: InterpResult<'tcx, ScalarMaybeUninit<Tag>>,
+        arg_type: &Ty<'tcx>,
+    ) -> InterpResult<'tcx, CArg> {
+        match scalar_res {
+            Ok(k) => {
+                // the ints
+                try_carg_from_prim_and_return!(hir::PrimTy::Int(IntTy::I8), k.to_i8(), CArg::Int8, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Int(IntTy::I16), k.to_i16(), CArg::Int16, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Int(IntTy::I32), k.to_i32(), CArg::Int32, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Int(IntTy::I64), k.to_i64(), CArg::Int64, arg_type);
+                // the uints
+                try_carg_from_prim_and_return!(hir::PrimTy::Uint(UintTy::U8), k.to_u8(), CArg::UInt8, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Uint(UintTy::U16), k.to_u16(), CArg::UInt16, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Uint(UintTy::U32), k.to_u32(), CArg::UInt32, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Uint(UintTy::U64), k.to_u64(), CArg::UInt64, arg_type);
+                // if no primitives were returned then we have an unsupported type
+                throw_unsup_format!(
+                    "Unsupported scalar argument type to external C function: {:?}",
+                    scalar_res
+                );
+            }
+            _ => {
+                throw_unsup_format!(
+                    "Unsupported argument type to external C function: {:?}",
+                    scalar_res
+                );
+            }
+        }
+    }
+
+    // call external C function and
+    // store output, depending on return type in the function signature
+    fn call_external_c_and_store_return<'a>(
+        &mut self,
+        external_fct_defn: ExternalCFuncDeclRep<'tcx>,
+        dest: &PlaceTy<'tcx, Tag>,
+        ptr: CodePtr,
+        libffi_args: Vec<libffi::high::Arg<'a>>,
+    ) -> InterpResult<'tcx, ()> {
+        let this = self.eval_context_mut();
+        unsafe {
```

Comment:
style nit: only wrap the actually unsafe part in `unsafe {}` (and document why the unsafe operation inside is sound at this site).